src/Sarif/FileRegionsCache.cs

```diff
@@ -47,24 +51,213 @@ public Region PopulatePrimaryRegionProperties(PhysicalLocation physicalLocation,
+            // If we have no input source file, there is no work to do
+            if (lineIndex == null) { return inputRegion; }
 
-            if (populateSnippet) { result.Snippet = new FileContent() { Text = "line1" }; }
+            Region region = inputRegion.DeepClone();
 
-            return result;
+            if (region.StartLine == 0)
+            {
+                // This means we have a region specified entirely via charOffset
+                PopulatePropertiesFromCharOffsetAndLength(lineIndex, region, fileText);
+            }
+            else
+            {
+                PopulatePropertiesFromStartAndEndProperties(lineIndex, region, fileText); 
+            }
+
+            if (populateSnippet)
+            {
+                region.Snippet = region.Snippet ?? new FileContent();
+
+                string snippetText = fileText.Substring(region.CharOffset, region.CharLength);
+                if (region.Snippet.Text == null)
+                {
+                    region.Snippet.Text = snippetText;
+                }
+                Debug.Assert(region.Snippet.Text == snippetText);
+            }
+
+            return region;
+        }
+
+        private void PopulatePropertiesFromCharOffsetAndLength(NewLineIndex newLineIndex, Region region, string fileText)
+        {
+            Debug.Assert(!region.IsBinaryRegion);
+            Debug.Assert(region.StartLine == 0);
+            Debug.Assert(region.CharLength > 0 || region.CharOffset > 0);
+
+            int startLine, startColumn, endLine, endColumn;
+
+            // Retrieve start and end line and column information from the new line index
+            OffsetInfo offsetInfo = newLineIndex.GetOffsetInfoForOffset(region.CharOffset);
+            startLine = offsetInfo.LineNumber;
+            startColumn = offsetInfo.ColumnNumber;
+
+            offsetInfo = newLineIndex.GetOffsetInfoForOffset(region.CharOffset + region.CharLength);
+            endLine = offsetInfo.LineNumber;
+
+            // The computation above points one past our actual region, because endColumn
+            // is exclusive of the region. This allows for length to easily be computed
+            // for single line regions: region.EndColumn - region.StartColumn
+            endColumn = offsetInfo.ColumnNumber;
+
+            // Only set values if they aren't already specified
+            if (region.StartLine == 0) { region.StartLine = startLine; }
+            if (region.StartColumn == 0) { region.StartColumn = startColumn; }
+            if (region.EndLine == 0) { region.EndLine = endLine; }
+            if (region.EndColumn == 0) { region.EndColumn = endColumn; }
+
+            // Validate cases where new line index disagrees with explicit values
+            Debug.Assert(region.StartLine == startLine);
+            Debug.Assert(region.StartColumn == startColumn);
+            Debug.Assert(region.EndLine == endLine);
+            Debug.Assert(region.EndColumn == endColumn);
```

Comment:
Zigackly. But in order for your validate to do that, we need this code to somehow flag the condition. Probably, we need a flag in this API, which would be 'overwrite any existing properties' or 'preserve existing properties'. the validator could call this code with 'overwrite any existing properties'. then it could send regions through this helper, compare the result to the region under validation, and fire any non-default properties that appear out of sync.

[https://github.com/oasis-tcs/sarif-spec/issues/206](https://github.com/oasis-tcs/sarif-spec/issues/206)

---
In reply to: [204582756](https://github.com/Microsoft/sarif-sdk/pull/957#discussion_r204582756) [](ancestors = 204582756)